### PR TITLE
Road to CoSWID Specification version 21

### LINKED ISSUE #14: Implement -17
Looking at https://tools.ietf.org/rfcdiff?url2=draft-ietf-sacm-coswid-17.txt it seems like we need to move around a number of smallish things:

* payload or evidence cardinality:
```
payload-or-evidence //= ( payload => payload-entry )
payload-or-evidence //= ( evidence => evidence-entry )
```

* ownership type (was unsigned):
```
$ownership /= int / text
```

* use type (was unsigned):
```
$use /= int / text	
```

* file-entry size type:
```
? size => uint,
```

* directory-entry path-elements is optional:
```
? path-elements => { path-elements-group },
```

* evidence-entry date type:
```
? date => integer-time,
```

* any-uri type
```
any-uri = uri
```

* 7.  Signed CoSWID Tags (currently not implemented)

* 8.  Tagged CoSWID Tags

### PR DESCRIPTION
**Pull request by CodingVoid**
Done by implementing the diff: https://tools.ietf.org/rfcdiff?url1=draft-ietf-sacm-coswid-16.txt&url2=draft-ietf-sacm-coswid-21.txt

- version-scheme type (was unsigned) is now signed
- ownership type (was unsigned) is now signed
- role type (was unsigned) is now signed
- use type (was unsigned) is now signed
- software-meta-entry->generator can now also be a 16 byte array
- file-entry global-attributes added
- directory-entry->path-elements is now optional:
- evidence-entry->location added:
- link-entry->rel (was unsigned) is now signed

breaking changes:
- ownership values changed (kind of swapped around...)
- filesystem-item->global-attributes removed
- There is now only a single payload or evidence (breaks JSON format, since it's always putting in an array). One could of course just leave it as it is and therefore stay compatible to CoSWID v16 and earlier, but I guess that's something to discuss.

potentially breaking changes (although unlikely):
- file-entry->size (was unsigned) is now signed
- payload and evidence are not arrays anymore, although I don't know I someone actually used multiple payloads or evicence's

CoSWID tagging and signing is still missing.

Fix #14

### CODE REVIEW

**Comment by thomas-fossati** [comment]
Great stuff and much needed, thanks @CodingVoid.  I'll have a look shortly.

File: common.go [hunk 1]
```diff
@@ -12,16 +12,25 @@ import (
...
 
 func stringifyCode(v *interface{}, dict codeDictionary, codeName string) error {
 	switch t := (*v).(type) {
 	case string:
 		*v = t
 		return nil
+	// uint64 is only for compatibility with code that uses this library and used
+	// the uint64 type before (as was default for CoSWID spec v16 and earlier)
 	case uint64:
+		if s, ok := dict[int64(t)]; ok {
```
Comment (thomas-fossati) [inline common.go:27]: I think we need to check that t is within 2⁶³-1 before converting.

File: common.go [hunk 2]
```diff
@@ -41,29 +50,38 @@ func codifyString(v *interface{}, dict stringDictionary) error {
 			*v = ui
 		}
 		return nil
+	// CBOR library returns uint64 type for positive integers, but we need int64
+	// for all types since CoSWID Spec v17
 	case uint64:
+		*v = int64(t)
```
Comment (thomas-fossati) [inline common.go:56]: ditto

File: link.go [hunk 3]
```diff
@@ -30,7 +31,7 @@ type Link struct {
 	//   For example, "swid:2df9de35-0aff-4a86-ace6-f7dddd1ade4c" references the
 	//   tag with the tag-id value "2df9de35-0aff-4a86-ace6-f7dddd1ade4c".
 	// * a URI with "swidpath:" as the scheme, which refers to another CoSIWD
-	//   via an XPATH query. This URI would need to be resolved in the context of
+	//   via an XPATH query [W3C.REC-xpath20-20101214] that matches items in that tag (Section 5.2). This URI would need to be resolved in the context of
```
Comment (thomas-fossati) [inline link.go:34]: missing line break

```suggestion
	//   via an XPATH query [W3C.REC-xpath20-20101214] that matches items in that
	//   tag (Section 5.2). This URI would need to be resolved in the context of
```

File: ownership.go [hunk 4]
```diff
@@ -14,28 +14,28 @@ type Ownership struct {
...
-   abandon=3
+   shared=3
 */
 
 // Ownership constants
 const (
-	OwnershipShared = uint64(iota + 1)
+	OwnershipShared = int64(iota + 1)
 	OwnershipPrivate
 	OwnershipAbandon
```
Comment (thomas-fossati) [inline ownership.go:27]: shouldn't we swap abandon and shared here?

**Review by thomas-fossati — COMMENTED** [review]
LGTM!  I've left a few comments inline.

File: common.go (hunk 2 above)
Comment (CodingVoid) [reply common.go:63]: done

File: ownership.go (hunk 4 above)
Comment (thomas-fossati) [reply ownership.go]: Note that changing this would make related tests in link_test.go fail.

**Review by thomas-fossati — APPROVED** [review]
amazing job, thanks a lot!

**Pull request merged**